### PR TITLE
Fix startup race condition when invoking the minibuffer.

### PR DESCRIPTION
**Pull request by Ambrevar**
Major change: `finalize' is now in charge for spawning the first window.
This should make `startup-function's much easier to write on the user end.

### CODE REVIEW

**Comment by Ambrevar** [comment]
@aartaka Can you confirm that you don't get the startup race condition you used to have?
@jmercouris Can you test on macOS?

**Comment by Ambrevar** [comment]
Great news don't come alone, and this fix might also fix the long (non-breaking) hang that we see on startup when many buffers get restored.

See #808 and #1012.

This hang seems to be due to us trying to restore too many buffers at the same time when initializing.  Now that we delay the restoration, it does not seem to hang the startup anymore.

**Comment by jmercouris** [comment]
And what if the user does not want to spawn a window?

**Comment by jmercouris** [comment]
Will they still be able to start Nyxt without spawning a window right away?

**Comment by jmercouris** [comment]
Actually you know what. I think you are right with your previous comment. The startup-function should be more limited in what it can do. If someone wants to dramatically change the startup of Nyxt it will take just a slight amount of effort more, but it will be a lot less effort for 99% of users who just want to change how their first buffer is formed.

File: source/browser.lisp [hunk 1]
```diff
@@ -212,10 +212,24 @@ editor executable."))
...
   (log4cl:add-appender log4cl:*root-logger* (make-instance 'messages-appender))
   (handler-case
-      (hooks:run-hook *after-init-hook*)
+      (hooks:run-hook *after-init-hook*) ; TODO: Run outside the main loop?
     (error (c)
       (log:error "In *after-init-hook*: ~a" c)))
-  (funcall-safely (startup-function browser) urls)
+  ;; Most of the code assumes the existence of a window.  Let's create it here
+  ;; so that the user does not have to bother with it in the `startup-function'.
+  (let ((window (window-make browser)))
```
Comment (jmercouris) [inline source/browser.lisp:220]: I disagree, I think the user should have to create the window in the startup-function.

**Comment by jmercouris** [comment]
Other than my comment about window creation, it looks OK to me! Thank you Pierre!

**Comment by Ambrevar** [comment]
I don't know if it makes sense to start Nyxt without starting a window right away.
If we want to support this use case nonetheless, then we need to redesign quite a few things in our code because too much of it assumes that a window exists.
So exposing an entry point to the user where a window does not exist is a recipe for disaster :p 

Anyways, in the context of this fix spawning the window beforehand matter: without it, there may be a race condition and the minibuffer would fail to spawn.

**Comment by Ambrevar** [comment]
Do not merge, I want to fix something else with the socket logic.

**Comment by Ambrevar** [comment]
I've just pushed a second commit which should help with the slow start when opening a URL externally.

The drawback is that `*socket-path*` can no longer be set from the init file.
It can be set from command line options, but obviously if the default browser is called by just `nyxt` it won't work.  So I guess we must introduce an environment variable here.  Thoughts?

**Comment by jmercouris** [comment]
What about checking for an environment variable, otherwise load the init file and try to contact Nyxt that way? That way if the environment variable is set, it will be faster, otherwise it will be the same speed as before

**Comment by jmercouris** [comment]
Scratch my above comment, we never did load the init file anyways, I e looked at your current implementation and it looks OK to me.

**Comment by Ambrevar** [comment]
Indeed, this fix does not change the behaviour, we've never allowed to set the
socket-path from the init file.

Even if we did, I think the strategy of loading the init file if the env
variable is not set would not solve the problem: if the init file takes,
say, 2s to load, then each URL opened externally will also take 2s to load.

**Comment by jmercouris** [comment]
Feel free to merge whenever :-)

**Comment by Ambrevar** [comment]
To be sure, have you tested on macOS?

**Comment by jmercouris** [comment]
I've not, I can do so

**Comment by jmercouris** [comment]
I can confirm that it works fine on macOS

**Comment by Ambrevar** [comment]
Thanks!  Will merge soon.

**Comment by aartaka** [comment]
> @aartaka Can you confirm that you don't get the startup race condition you used to have?

I don't get the race condition. Works as a charm!

**Comment by Ambrevar** [comment]
Thanks for reporting!


There are news however: I'm trying to fix the libsoup collision bug once and for
all.
The easiest way to fix it is to avoid creating web-buffer web view at the same
time.

9094ff2f489ef514e7d587f1abcde4309f4aa5b5 fixes the startup collisions by not
making contexts for internal buffers.  This is a great improvement already!

(Feedback welcome!)

Now a little bit harder is to fix the collisions that happen when restoring a
session.

Commit 168e5bd587ce434214b623a28d7c1e70f59f9537 adds a 0.05s sleep between each
buffer, which is enough on my machine but might not work on slower systems.
This is obviously very brittle.

A cleaner fix would be to separate buffer create from web view creation, and
only create the latter on demand, e.g. when we call window-set-current-buffer.

This has some significant implications though.  For instance `buffer-make`
would not initialize the modes anymore, the web-view creation function should do
it instead.

Before going down this rabbit hole, I'd like to confirm with you that this is a
sound decision.

Thoughts? @jmercouris @aartaka?

**Comment by jmercouris** [comment]
It is most probably not a sound decision :-D. I would suggest to remove this fix from this PR so that we can merge it and master works again.

I would hold off on trying to patch the behavior of GTK bugs in Lisp. I suggest that we implement gobject backend before spending any time trying to fix this bug. If we are to fix this bug, we should make it transparent to our processes, the API implementation (aka renderer-gtk.lisp) should handle all details. There should be no special cases, if there are we need to improve the API or the API implementation.

**Comment by Ambrevar** [comment]
John Mercouris <notifications@github.com> writes:

> I would suggest to remove this fix from this PR so that we can merge it and master works again.

Why remove the fix?  At worst, the fix does not change anything on
slower hardware.  At best, it makes for a smoother experience on fast hardware.

> I would hold off on trying to patch the behavior of GTK bugs in
> Lisp. I suggest that we implement gobject backend before spending any
> time trying to fix this bug.

This bug is in Libsoup, see

https://gitlab.gnome.org/GNOME/libsoup/-/issues/192

It'll be there with the gobject backend as well.

> If we are to fix this bug, we should make it transparent to our processes, the API implementation (aka renderer-gtk.lisp) should handle all details. There should be no special cases, if there are we need to improve the API or the API implementation.

This is a fair point.

My suggestion of having just-in-time webview instantiation could be
restricted to renderer-gtk I believe.  I'll work out the details later.

**Comment by Ambrevar** [comment]
I've tried 2 different approaches:

- Instantiate the buffer gtk-object in the accessor.
  This alone does not work change anything because when we initialize a buffer,
  we initialize the modes, which in turn may need access to the gtk-object
  (proxy-mode does).

  The way forward would be to delay mode initialization to when gtk-object is
  initialized.

  This does not seem well abstracted because now we have to deal with
  half-initialized buffers and it's unclear to me whether all the code base can
  properly deal with that.  We lack a single entry point for buffer initialization.

- Add the history entries directly to the buffer list.
  Then in `window-set-active-buffer` check if we are given a buffer or a
  history.  If a history, instantiate the corresponding buffer, when set this
  buffer.

  The problem is that now `buffer-suggestion-filter` needs to deal with both
  buffers and histories.  In general, any piece of code that relies on
  `buffer-list` would need to be updated.
  
Another approach which I haven't tried, but probably cleaner than the above.

- Define an unrestored-gtk-buffer class between `buffer` and `gtk-buffer`.
  This class would not have an `initialize-instance` after method.

  Then the session `restore` method would fill the buffer list with this
  unrestored-gtk-buffers which can be instantiated to full-blown `gtk-buffer`s
  just in time.

  We would still need a entry point to buffer initialization and the above still
  has the problem that our code base now would need to differentiate between
  gtk-buffers and unrestored-gtk-buffers.

- Better: instead of adding the unrestored buffers to `buffers`, add the session
  histories to a new slot, "unrestored buffers".  Then with the new minibuffer
  multi-source support, we can list these entries alongside the other buffers.
  This would give us clean data separation and not break any of the existing code.

This raises a question about the similarity between dead buffers (killed, but
can brought back to life with `reopen-buffer` and unrestored buffers.  Can both
be conflated?  I'm not sure because unrestored buffer only come with a history
and that's it, while dead buffers still hold the mode instances in memory.

**Comment by Ambrevar** [comment]
Except for the sleep in the last commit, I've merged this as 4a1beacbe37014b542f5783811d9e681e36d9443.

**Comment by Ambrevar** [comment]
We can keep discussing lazy buffer restoration in #1076.

**Pull request closed without merging**